File: leetcode/0047/permuteUnique.swift
var n: Int = 0
var list: [Int] = []
var result = [[Int]]()
var used = [Bool]()

func permuteUnique(_ nums: [Int]) -> [[Int]] {
    list = nums.sorted()
    n = nums.count
    
    used = Array.init(repeating: false, count: n)
    backtrace(depath: 0, path: [])
    return result
}

func backtrace(depath: Int = 0, path: [Int]) {
    if depath == n {
        result.append(path)
    } else {
        var path = path
        for i in 0..<n {
            if used[i] || (i > 0 && list[i] == list[i-1] && !used[i-1]) {
                continue
            }
            used[i] = true
            path.append(list[i])
            backtrace(depath: depath + 1, path: path)
            used[i] = false
            path.popLast()
        }
    }
}

print(permuteUnique([1,1,2]))

File: leetcode/0047/permuteUnique.py
def permuteUnique(nums: [int]) -> [[int]]:
    def dfs(depth, path):
        if depth == size:
            res.append(path.copy())
            return
        for i in range(size):
            if used[i] or (i > 0 and nums[i] == nums[i - 1] and not used[i - 1]):
                continue

            used[i] = True
            path.append(nums[i])
            dfs(depth + 1, path)
            used[i] = False
            path.pop()

    size = len(nums)
    if size == 0: return []
    nums.sort()

    used = [False] * len(nums)
    res = []
    dfs(0, [])
    return res

print(permuteUnique([1,1,2]))